plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/PublishDiagnosticsProcessor.java

```diff
@@ -12,29 +12,61 @@
 
 import com.google.inject.Inject;
 import com.google.inject.Singleton;
+import java.util.List;
+import java.util.Optional;
 import org.eclipse.che.api.languageserver.shared.model.ExtendedPublishDiagnosticsParams;
+import org.eclipse.che.ide.api.data.tree.Node;
 import org.eclipse.che.ide.api.editor.EditorAgent;
 import org.eclipse.che.ide.api.editor.EditorPartPresenter;
 import org.eclipse.che.ide.api.editor.annotation.AnnotationModel;
 import org.eclipse.che.ide.api.editor.editorconfig.TextEditorConfiguration;
 import org.eclipse.che.ide.api.editor.texteditor.TextEditor;
+import org.eclipse.che.ide.part.explorer.project.ProjectExplorerPresenter;
 import org.eclipse.che.ide.resource.Path;
+import org.eclipse.che.ide.resources.tree.ResourceNode;
+import org.eclipse.che.ide.ui.smartTree.Tree;
 import org.eclipse.lsp4j.Diagnostic;
 
 /** @author Anatolii Bazko */
 @Singleton
 public class PublishDiagnosticsProcessor {
 
   private final EditorAgent editorAgent;
+  private final Tree tree;
 
   @Inject
-  public PublishDiagnosticsProcessor(EditorAgent editorAgent) {
+  public PublishDiagnosticsProcessor(
+      EditorAgent editorAgent, ProjectExplorerPresenter projectExplorer) {
     this.editorAgent = editorAgent;
+    this.tree = projectExplorer.getTree();
   }
 
-  public void processDiagnostics(ExtendedPublishDiagnosticsParams diagnosticsMessage) {
-    EditorPartPresenter openedEditor =
-        editorAgent.getOpenedEditor(new Path(diagnosticsMessage.getParams().getUri()));
+  public void processDiagnostics(List<ExtendedPublishDiagnosticsParams> diagnostics) {
+    tree.getNodeStorage()
+        .getAll()
+        .stream()
+        .filter(node -> node instanceof ResourceNode)
+        .forEach(
```

Comment:
Hi,

I guess we can add something like `.map(ResourceNode.class::cast)` here to make `forEach` section more concise.